frontend/src/Grid.tsx

```diff
@@ -0,0 +1,201 @@
+import React, { useEffect, useCallback, useState } from "react";
+import "./Grid.css";
+import "./Crossword.css";
+
+import { WriteMode } from "./Crossword";
+
+// Define array type that only accepts a specified number of elements
+// https://stackoverflow.com/a/59906630
+type ArrayLengthMutationKeys = 'splice' | 'push' | 'pop' | 'shift' | 'unshift' | number
+type ArrayItems<T extends Array<any>> = T extends Array<infer TItems> ? TItems : never
+type FixedLengthArray<T extends any[]> =
+  Pick<T, Exclude<keyof T, ArrayLengthMutationKeys>>
+  & { [Symbol.iterator]: () => IterableIterator<ArrayItems<T>> }
+
+export type Vec2 = FixedLengthArray<[number, number]>
+export type Vec3 = FixedLengthArray<[number, number, number]>
+export type Vec4 = FixedLengthArray<[number, number, number, number]>
+
+interface GridProps {
+  squares: {
+    medianLen: number,
+    grid: Square[][],
+  };
+  letters: LetterType[];
+  cursorPosition: Vec2;
```

Comment:
Why not just `[number, number]`. Haven't written typescript (nor js for that matter) in years :)

There is no input validation anywhere anyways.

Reply:
Because nothing in typescript restricts you from doing, for example,
```
let a: [number, number] = [0, 1]
a.push(2)
```
And now we have "ruined" our cursor position :P. Also, the `Vec2` in this implementation does not let you pick an invalid index
```
const b: Vec2 = [0, 1]
const c: number = b[2] // throws error
```
(https://stackoverflow.com/a/59906630)